tools/topology/topology1/sof-tgl-rt1308-hdmi-ssp.m4

```diff
@@ -0,0 +1,217 @@
+
+# capture DAI using 2 periods
+# Buffers use s16le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	2, SSP, HDMI_2_SSP_NUM, HDMI2_SSP_NAME,
+	PIPELINE_SINK_2, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+ifdef(`NO_AMP',`',`
+# playback DAI using 2 periods
+# Buffers use s24le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	3, SSP, AMP_SSP_NUM, AMP_SSP_NAME,
+	PIPELINE_SOURCE_3, 2, s24le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)')
+
+# capture DAI is DMIC01 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	4, DMIC, 0, dmic01,
+	PIPELINE_SINK_4, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp1 using 2 periods
+# Buffers use s32le format, 1000us deadline with priority 0 on core 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+        5, HDA, 5, iDisp1,
+        PIPELINE_SOURCE_5, 2, s32le,
+        1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp2 using 2 periods
+# Buffers use s32le format, 1000us deadline with priority 0 on core 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+        6, HDA, 6, iDisp2,
+        PIPELINE_SOURCE_6, 2, s32le,
+        1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp3 using 2 periods
+# Buffers use s32le format, 1000us deadline with priority 0 on core 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+        7, HDA, 7, iDisp3,
+        PIPELINE_SOURCE_7, 2, s32le,
+        1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# PCM Low Latency, id 0
+dnl PCM_CAPTURE_ADD(name, pcm_id, capture)
+PCM_CAPTURE_ADD(HDMI-IN-1, 0, PIPELINE_PCM_1)
+
+dnl PCM_CAPTURE_ADD(name, pcm_id, capture)
+PCM_CAPTURE_ADD(HDMI-IN-2, 1, PIPELINE_PCM_2)
+
+ifdef(`NO_AMP',`',`
+dnl PCM_PLAYBACK_ADD(name, pcm_id, playback)
+PCM_PLAYBACK_ADD(Speaker, 2, PIPELINE_PCM_3)')
+PCM_CAPTURE_ADD(DMIC01, 3, PIPELINE_PCM_4)
+
+PCM_PLAYBACK_ADD(HDMI 1, 5, PIPELINE_PCM_5)
+PCM_PLAYBACK_ADD(HDMI 2, 6, PIPELINE_PCM_6)
+PCM_PLAYBACK_ADD(HDMI 3, 7, PIPELINE_PCM_7)
+
+#
+# BE configurations - overrides config in ACPI if present
+#
+
+#HDMI-1 SSP (ID: 0)
+DAI_CONFIG(SSP, HDMI_1_SSP_NUM, 0, HDMI1_SSP_NAME,
+	   SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
+		      SSP_CLOCK(bclk, 3072000, codec_master),
+		      SSP_CLOCK(fsync, 48000, codec_master),
+		      SSP_TDM(2, 32, 3, 3),
+		      SSP_CONFIG_DATA(SSP, HDMI_1_SSP_NUM, 32, 0)))
+
+#HDMI-2 SSP (ID: 1)
+DAI_CONFIG(SSP, HDMI_2_SSP_NUM, 1, HDMI2_SSP_NAME,
+	   SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
+		      SSP_CLOCK(bclk, 3072000, codec_master),
+		      SSP_CLOCK(fsync, 48000, codec_master),
+		      SSP_TDM(2, 32, 3, 3),
+		      SSP_CONFIG_DATA(SSP, HDMI_2_SSP_NUM, 32, 0)))
+
+ifdef(`NO_AMP',`',`
+#Amplifier Codec SSP (ID: 2)
+DAI_CONFIG(SSP, AMP_SSP_NUM, 2, AMP_SSP_NAME,
+	SSP_CONFIG(I2S, SSP_CLOCK(mclk, 38400000, codec_mclk_in),
```

Comment:
that doesn't seem right, why would you receive the mclk and provide the bclk/fsync to the codec?

Is this intentional and if yes can you add a comment?

Reply:
Yes. Amplifier case still its required and used.  For NO_AMP boards, this config wont be enabled.

Reply:
But you are using two different MCLK in config, with different values. How is this even supposed to work?

Reply:
Right. HDMI-In capture case really the MCLK is not used Since Lontium chip is master and DSP config not required. But in codec amplifier case it has been used our platform (TGL) MCLK (38.4 Mhz)configuration and value.

Reply:
Hi plbossart,

Hope I have clarified your question. Please help to proceed further if you have no more question.

Reply:
I don't see anything new to review @CBala21

Reply:
> I don't see anything new to review @CBala21

Could you provide your approval? waiting for one more approval.

Reply:
I am not convinced your solution works, the two MCLK configuration will conflict. To be clearer, the MCLK is unrelated to SSP, if you configure the MCLK in different ways for different SSPs, it's still the SAME MCLK hardware.

Reply:
Ok. I meant that HDMI-in capture case really NO MCLK required from platform Hardware and its generated and provided by external Lontium card which is converting HDMI inputs into I2S PCM and sending PCM samples via I2S pins. Hence what ever the values we mentioned in topology will not impact and not programmed by our DSP because we used config flag as "codec_master". Assumed in this case our DSP not consider MCLK value and wont generate and program MCLK. only bclk/fsync would be considered to support 32 bit and 48 Khzsamples.

also as my understanding, for codec amplifier case which ever the MCLK used is important to operate SSP amplifier. since we used flag "codec_slave" and DSP would program it.
we just followed the other codec amplifier topologies as a reference.

Please correct me if am wrong and suggest if any expectation on the values/parameters used on DAI CONFIG for HDMI-In capture. by the way the use case has been tested and delivered to TGL customers long time before and no issue found till now.

Reply:
I don't think we have a macro to say 'NO MCLK'. @ranj063 may need to correct me here.

What you could do is use the same settings as for the amplifier, and add a comment that it's not used for the HDMI input. That way you wouldn't have any issue with conflicts on the MCLK setting.

We've done this before and used the same MCLK frequency/direction in other cases.

Reply:
@plbossart  @CBala21 Petér has been working on improving handling this case in https://github.com/thesofproject/sof/pull/7741 ... i.e. when SSP is not clock provider, we do not need to program MCLK. This was discovered in similar case where multiple SSPs were running with conflicting MCLKs  -> https://github.com/thesofproject/sof/issues/7803

Reply:
> I don't think we have a macro to say 'NO MCLK'. @ranj063 may need to correct me here.
> 
> What you could do is use the same settings as for the amplifier, and add a comment that it's not used for the HDMI input. That way you wouldn't have any issue with conflicts on the MCLK setting.
> 
> We've done this before and used the same MCLK frequency/direction in other cases.

Updated the file as per your suggestion. please check and provide the feedback.